# Neural Networks

Explication : Réseaux de Neurones pour la Reconnaissance des Chiffres Manuscrits

Les réseaux de neurones artificiels (ANN, pour Artificial Neural Networks) sont des modèles d'apprentissage supervisé inspirés du cerveau humain. Ils sont particulièrement efficaces pour traiter des données complexes comme des images, du texte ou des sons. Voici une explication complète sur leur fonctionnement et leur utilisation dans ce projet.


1. Compréhension des Réseaux de Neurones
-Entrée (Input) : Chaque réseau prend des caractéristiques en entrée, comme les pixels d’une image. Dans ce projet, les images des chiffres manuscrits sont de taille 28×2828×28 pixels, ce qui donne 784 caractéristiques en entrée.

-Couches Cachées (Hidden Layers) : Des couches de neurones appliquent des transformations mathématiques pour capturer les relations complexes entre les données d'entrée et la sortie.

-Sortie (Output) : La dernière couche donne les résultats sous forme de probabilités pour chaque classe. Ici, il y a 10 sorties, correspondant aux chiffres de 0 à 9.

-Apprentissage : En utilisant les données d’entraînement, le modèle ajuste les poids des connexions entre les neurones pour minimiser l’erreur entre la sortie prédite et la sortie réelle.

Projet de reconnaissance les chiffres manuscrits avec les reseaux de neuronnes

In [10]:
import cv2 as cv
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf

# 2. Structure du Code

##    Chargement des Données :

In [11]:
mnist = tf.keras.datasets.mnist

In [12]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()

11490434/11490434 [==============================] - 1s 0us/step


tf.keras.datasets.mnist : Base de données contenant des images de chiffres manuscrits (60,000 images pour l’entraînement et 10,000 pour le test).
x_train et x_test : Contiennent les images des chiffres.
y_train et y_test : Contiennent les étiquettes correspondantes (valeurs entre 0 et 9).

## Prétraitement des Données :

tf.keras.datasets.mnist : Base de données contenant des images de chiffres manuscrits (60,000 images pour l’entraînement et 10,000 pour le test).
x_train et x_test : Contiennent les images des chiffres.
y_train et y_test : Contiennent les étiquettes correspondantes (valeurs entre 0 et 9).

## Prétraitement des Données :

In [13]:
x_train = tf.keras.utils.normalize(x_train, axis=1)
x_test = tf.keras.utils.normalize(x_test, axis=1)

Les pixels des images sont normalisés pour être compris entre 0 et 1 (au lieu de 0 à 255). Cela accélère et améliore l'apprentissage.

# Création du Modèle :

In [14]:
model = tf.keras.models.Sequential()

## Ajout des Couches :

In [17]:
model.add(tf.keras.layers.Flatten(input_shape=(28,28)))
model.add(tf.keras.layers.Dense(units=128, activation=tf.nn.relu))
model.add(tf.keras.layers.Dense(units=128, activation=tf.nn.relu))
model.add(tf.keras.layers.Dense(units=10, activation=tf.nn.softmax))

Flatten : Transforme une image 28×2828×28 en un vecteur de 784 valeurs.
Couches Denses :

    Dense(128, activation='relu') : Ajoute une couche de 128 neurones avec une activation ReLU (Rectified Linear Unit). Cela permet au modèle de capturer des relations non linéaires.
    Dense(10, activation='softmax') : La dernière couche contient 10 neurones, chaque neurone représentant un chiffre (0 à 9). Softmax transforme les valeurs en probabilités.

# Compilation et Entraînement :

In [21]:
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

In [20]:
model.fit(x_train, y_train, epochs=3)

Epoch 1/3
1875/1875 [==============================] - 4s 2ms/step - loss: 4.3401 - accuracy: 0.1965
Epoch 2/3
1875/1875 [==============================] - 4s 2ms/step - loss: 4.1799 - accuracy: 0.2035
Epoch 3/3
1875/1875 [==============================] - 4s 2ms/step - loss: 4.1174 - accuracy: 0.2092


Compilation :

    optimizer='adam' : Algorithme d'optimisation efficace pour ajuster les poids.
    loss='sparse_categorical_crossentropy' : Fonction de perte utilisée pour les problèmes de classification multiclasse.
    metrics=['accuracy'] : Évalue la précision du modèle.

Entraînement : Le modèle est entraîné sur les données d’entraînement pendant 3 époques (itérations complètes sur les données).

# Évaluation du Modèle :

In [24]:
loss, accuracy = model.evaluate(x_test, y_test)
print(accuracy)
print(loss)

313/313 [==============================] - 1s 2ms/step - loss: 4.4532 - accuracy: 0.2352
0.23520000278949738
4.453223705291748


Évalue la performance sur l’ensemble de test et affiche la précision et la perte.

# Sauvegarde et Prédiction :

In [25]:
model.save('digits.model')

INFO:tensorflow:Assets written to: digits.model\assets


Le modèle est sauvegardé pour être utilisé ultérieurement.

Prédiction sur de nouvelles images :

for x in range(1,6):
    img = cv.imread(f'{x}.png')[:,:,0]
    img = np.invert(np.array([img]))
    prediction = model.predict(img)
    print(np.argmax(prediction))
    plt.imshow(img[0], cmap=plt.cm.binary)
    plt.show()

Les images sont chargées, inversées (pixels noirs/blancs) et prétraitées avant d’être passées au modèle pour prédiction.
np.argmax(prediction) : Retourne la classe (chiffre) avec la probabilité la plus élevée.

# Résultats

    Précision sur les données de test : 23.5%23.5%
    Perte sur les données de test : 4.4534.453
    La faible précision peut indiquer que le modèle nécessite plus d’entraînement ou que les hyperparamètres doivent être ajustés (par exemple, plus d'époques, couches supplémentaires, etc.).

# Améliorations Potentielles

-Augmenter les Époques : Entraîner le modèle pendant plus de 3 époques peut améliorer la précision.

-Augmentation des Données : Appliquer des transformations (rotation, zoom, etc.) pour augmenter les données d’entraînement.

-Hyperparamètres :
    -Ajuster la taille des couches (nombre de neurones).
    -Utiliser d’autres fonctions d’activation ou algorithmes d’optimisation.

-Convolutional Neural Networks (CNN) : Remplacer ce modèle dense par un réseau de neurones convolutifs, spécialement conçu pour les images.

Conclusion

Ce projet montre comment les réseaux de neurones peuvent être utilisés pour reconnaître des chiffres manuscrits. En combinant prétraitement des données, construction d’un modèle neuronal, et ajustement des hyperparamètres, il est possible de créer un modèle performant pour une tâche donnée. Les résultats peuvent être affinés en explorant des techniques avancées comme les CNN.